### ISSUE TITLE
Support static file/asset handling (aka file-loader/url-loader in webpack)

### ISSUE THREAD

**Issue by swernerx** [rollup/rollup#861]
Using webpack it's possible to define requirements of e.g. a component with imports to static files/assets e.g. fonts, css files, images. I would like to bundle code from a UI library project using rollup but therefor it has to keep the imports for these static file assets in-tact so that these are usable by webpack. Through the fact that bundling also changes relative paths it has to copy references files to the "lib" or "dist" folder and adjust these references.

I tried implementing this. What I am currently working on is a node-based [utility for simplifying pre-publishing of libraries](https://github.com/sebastian-software/prepublish-lib) so that these work in the different destination environments. The [rollup-plugin](https://github.com/sebastian-software/prepublish-lib/blob/master/src/loader.js) is currently integrated in the tool, but I would definitely extract it into a separate solution/npm package once I feel that there is a solution. What I basically need to make the remaining issues fixed is that I can write out "import" statements which are not further processed by rollup.

My original approach was to just implement the `load(id)` method. There I am actually taking care of copying the source files to the destination folder and returning a valid ES modules import to the new CSS file. Unfortunately this does not work yet as rollup tries to then find the copied CSS file relative to the original file which does not work. What I miss is a possibility to write out an 'import' statement which is ignored by rollup in the later processing. Something which is comparable to the exclusion of `node_modules` when the required plugins are not available.

**Comment by swernerx** [comment]
I made some progress. Actually I figure I have to implement `resolveId` for filtering the "new" copied over files. The documentation mentions that returning falsy values will mark them as external. In my experiments this does not seem to work.

Output:

```
Processing style: /Users/Sebastian/Workspace/sebastian-software/prepublish-lib/test/index.css
Processed: /Users/Sebastian/Workspace/sebastian-software/prepublish-lib/test/index.css => lib/index-6abc46bc.css
RR: Testing: ./index-6abc46bc.css
RR: Marking external!
Error: Could not resolve './index-6abc46bc.css' from /Users/Sebastian/Workspace/sebastian-software/prepublish-lib/test/index.css
    at /Users/Sebastian/Workspace/sebastian-software/prepublish-lib/node_modules/rollup/dist/rollup.js:7416:40
```

Relevant Code Block:

```js
    resolveId: function(importee, importer) {
      console.log("RR: Testing:", importee)
      if (importee in externalIds) {
        console.log("RR: Marking external!")

        // This does not seem to work!

        // "returning any other falsy value signals that importee should be treated as an external module
        // and not included in the bundle." -- https://github.com/rollup/rollup/wiki/Plugins#creating-plugins
        return false
      }

      console.log("RR: Delegating...")
    },
```

`externalIds` is a map which registers all the dynamically copied files. The lookup there seems to work when looking at the output.

**Comment by swernerx** [comment]
BTW: rollup version 0.34.9

**Comment by Swatinem** [comment]
You might want to take a look at https://github.com/Swatinem/rollup-plugin-url
Although I’m not really maintaining it anymore since I use something different now. Possible improvements are to use the new rollup builtin onwrite hook.

**Comment by swernerx** [comment]
Thanks for sharing. That's indeed quite interesting. The "write" method which you use does not seem to exist anymore... or is not documented in the rollup wiki. 

I don't understand how it works for a non-inlining scenario. In this case it seems to produce code like "export default file_hash.ext" which actually exports just the filename. Just exporting a file name does not add the file as a dependency to the original requesting file. 

This differs quite a bit from the thing I need. I'd like to keep the import statement in-tact and just update the file reference.

I'd like to use the exported project (library) inside another project (app) which uses webpack v2 for building.

Any docs on the `onwrite` hook and what it does?

**Comment by Rich-Harris** [comment]
I was going to suggest something similar to @Swatinem – if you had something like this...

```
// app.js
import audioFileUrl from '../assets/audio/somefile.mp3';

const audio = new Audio();
audio.src = audioFileUrl; // e.g. `/assets/audio/somefile.${hash}.mp3`
audio.play();
```

...then the plugin would find `somefile.mp3`, optionally hash the filename, copy it to a specified destination directory (or do the data URI thing), and generate a module exporting a string containing a URL. In other words, `../assets/audio/somefile.mp3` *shouldn't* be treated as an external module.

I think rollup-plugin-url is most of the way there, though I couldn't see a way to generate a URL that was different from the path on the filesystem (e.g. I might want the plugin to copy the file to a `build` folder, but I don't want that to be in the generated URL) – perhaps I'm missing something?

Also, because of the incremental caching stuff we added recently it's much better if plugins are stateless, i.e. no objects like `copies` – copying should happen straight away. I think we would probably also want the copying to happen inside `transform` rather than `load`, as counter-intuitive as that sounds, because then files won't be recopied on incremental rebuilds unless they've changed.

So in summary I think we're almost there, and that we should focus our efforts on getting rollup-plugin-url to do everything we need. Great work @Swatinem 😀 – if you don't have plans to maintain it, would you be happy adding me as a contributor or transferring it to the rollup organisation so we can pick up where you left off?

**Comment by Rich-Harris** [comment]
@swernerx Ha, timing!

**Comment by swernerx** [comment]
As said IMHO what's missing from is effectively keeping the import statement. For later processing in another project via Webpack + File-Loader I don't need the URL as a string but a real es2015 import statement.

**Comment by Swatinem** [comment]
I’m happy to transfer ownership. I invited you as collaborator, will make you admin once you accept. Seems like I need admin privileges on the `rollup` org to transfer it myself.

**Comment by swernerx** [comment]
- File content inlined as generated by 'rollup-plugin-url`: `"file_hash.ext"`
- Required for Webpack file-loader or any other follow-up processing: `import path_to/file_hash.ext`

**Comment by Swatinem** [comment]
Oh, I see.
@swernerx The goal of this plugin was basically to do everything that webpacks file-loader/url-loader were doing.
Like copy the file to a hashed filename inside the output directory.

So when you want to do the loading/copying via webpack in a later step, maybe you would like to just flat out ignore the file so that rollup will just keep the import as-is.

**Comment by swernerx** [comment]
Yep, that's exactly what I tried to implement with returning resolveId(hashedFile) => false - which as I said does not seem to work.

**Comment by swernerx** [comment]
I see that the naming/title is probably wrong: I was not looking for a file-loader/url-loader requirement but for a solution for flatten required files into an output/dist directory and making the bundle aware of the new locations.

I could either be able to use the output with @Swatinem plugin or Webpack's file/url loaders.

**Comment by swernerx** [comment]
My ideal output would be something which produces:

- bundle.js
- asset1.png
- font1.ttf
- etc.

all in one directory. Flattened. And possibly named by their content hash.

**Comment by swernerx** [comment]
- rollup-url-loader: app level replacement for webpacks loaders
- my-loader: library level prepublish tool for bundling all js with their assets into one folder (for publishing to npm in a follow-up step)

**Comment by Rich-Harris** [comment]
@Swatinem thanks

@swernerx Ah, ok – so the nub of the issue is about how to make `resolveId` behave correctly. Not entirely sure what's going on there, will look into it. A workaround is to have an `options.external` function that filters out the static asset file paths as they're generated. Demo here – it's a little bit messy but it works: https://github.com/Rich-Harris/rollup-issue-861

**Comment by swernerx** [comment]
@Rich-Harris this seems to work. Only issue is that instead of the `importee, importer` combination of `resolveId` it's using the full file path.

```js
    var fileMapper = loader(outputFolder)

    return rollup({
      ...
      external: function(id) {
        return fileMapper.isExternal(id)
      },
      ...
   })
```

and this method inside the plugin:

```js
    isExternal: function(id) {
      var baseName = "./" + path.basename(id)
      if (baseName in externalIds) {
        return true;
      }
    }
```

No I am just getting another quite confusing message:

`No name was provided for external module '/Users/Sebastian/Workspace/sebastian-software/prepublish-lib/test/index-6abc46bc.css' in options.globals – guessing 'h6abc46bc'`

**Comment by swernerx** [comment]
See also the commit to my prepublish tool: https://github.com/sebastian-software/prepublish-lib/commit/cbe4872acd303319ab0cd48a889212f7cae6a06a

**Comment by Rich-Harris** [comment]
> No I am just getting another quite confusing message:

You'll see that if you're generating a UMD or IIFE bundle – external modules need to exist on the page as globals. If you're generating an AMD, CJS or ES bundle you won't see that (and if you're generating a bundle that Webpack will later consume, presumably you want either ES or CJS?)

**Comment by swernerx** [comment]
I do both ES and CJS. But also UMD for direct consumption inside the browser. I know that possibly not all features makes sense for that format. But why is an only internal known variable even relevant. The name must only work inside the closure and is not "published" to the outside in this case, right?

**Comment by swernerx** [comment]
I published my solution to NPM and added an entry to this wiki page: https://github.com/rollup/rollup/wiki/Plugins

I called the plugin "relink" as I figured it's a good match for what it actually does. Hopefully it's useful for other people. It actually does a little bit more using some [PostCSS](http://postcss.org) magic where it includes all included CSS files to produce flat output stylesheets.

I use it as a part of my library publishing solution [prepublish-lib](https://github.com/sebastian-software/prepublish-lib) which is an easy to use tool for building the typical publish-ready output formats (cjs + es2015 modules) from the source code.

**Comment by TrySound** [comment]
Thanks

**Comment by Artur93gev** [comment]
@swernerx could pls provide an information does your solution working now?
Im facing the issue with `rollup-rebase`((
Any help will be great

**Comment by bionicles** [comment]
would use m4a files to play annoying fingersnap sound when users add nodes and links in a graph db. anyone had luck with audio?
